nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_common/connections/udp.hpp

```diff
@@ -0,0 +1,383 @@
+      while (state_ == State::ACTIVE) {
+        auto data_available = is_data_available();
+        if (!data_available.has_value()) throw SocketError(data_available.error());
+        if (!data_available.value()) continue;
+
+        buffer.resize(buffer_size_);
+        auto msg_header = make_msg_header(buffer);
+
+        ssize_t recv_result = recvmsg(sock_fd_, &msg_header.msg, MSG_TRUNC);
+        if (recv_result < 0) throw SocketError(errno);
+        size_t untruncated_packet_length = recv_result;
+
+        if (!is_accepted_sender(msg_header.sender_addr)) continue;
+
+        ReceiveMetadata metadata;
+        get_receive_metadata(msg_header.msg, metadata);
+        metadata.truncated = untruncated_packet_length > buffer_size_;
+
+        buffer.resize(std::min(buffer_size_, untruncated_packet_length));
+        callback_(buffer, metadata);
+      }
+    });
+  }
+
+  void get_receive_metadata(msghdr & msg, ReceiveMetadata & inout_metadata)
+  {
+    for (cmsghdr * cmsg = CMSG_FIRSTHDR(&msg); cmsg != nullptr; cmsg = CMSG_NXTHDR(&msg, cmsg)) {
+      if (cmsg->cmsg_level != SOL_SOCKET) continue;
+
+      switch (cmsg->cmsg_type) {
+        case SO_TIMESTAMP: {
+          auto tv = (timeval const *)CMSG_DATA(cmsg);
+          uint64_t timestamp_ns = tv->tv_sec * 1'000'000'000 + tv->tv_usec * 1000;
+          inout_metadata.timestamp_ns.emplace(timestamp_ns);
+          break;
+        }
+        case SO_RXQ_OVFL: {
+          auto drops = (uint32_t const *)CMSG_DATA(cmsg);
+          inout_metadata.drops_since_last_receive =
+            drop_monitor_.get_drops_since_last_receive(*drops);
+          break;
+        }
+        default:
+          continue;
+      }
+    }
+  }
+
+  util::expected<bool, int> is_data_available()
+  {
+    int status = poll(&poll_fd_, 1, g_poll_timeout_ms);
+    if (status < 0) return errno;
+    return (poll_fd_.revents & POLLIN) && (status > 0);
+  }
+
+  bool is_accepted_sender(const sockaddr_in & sender_addr)
+  {
+    if (!sender_) return true;
+
+    std::array<char, INET_ADDRSTRLEN> sender_name;
+    inet_ntop(AF_INET, &sender_addr.sin_addr, sender_name.data(), INET_ADDRSTRLEN);
+    return std::strncmp(sender_->ip.c_str(), sender_name.data(), INET_ADDRSTRLEN) == 0;
+  }
+
+  MsgBuffers make_msg_header(std::vector<uint8_t> & receive_buffer) const
+  {
+    msghdr msg{};
+    iovec iov{};
+    std::array<std::byte, 1024> control;
+
+    sockaddr_in sender_addr;
+    socklen_t sender_addr_len = sizeof(sender_addr);
+
+    iov.iov_base = receive_buffer.data();
+    iov.iov_len = receive_buffer.size();
+
+    msg.msg_iov = &iov;
+    msg.msg_iovlen = 1;
+    msg.msg_control = control.data();
+    msg.msg_controllen = control.size();
+    msg.msg_name = &sender_addr;
+    msg.msg_namelen = sender_addr_len;
+
+    return MsgBuffers{msg, iov, control, sender_addr};
+  }
+
+  std::atomic<State> state_{State::UNINITIALIZED};
+
+  int sock_fd_;
+  pollfd poll_fd_;
+
+  size_t buffer_size_ = 1500;
```

Comment:
uniform initialization and const/constexpr for the 1500?

Reply:
It can be changed during runtime via `set_mtu`, but I agree with uniform initialization.